core/consensus/src/engine.rs

```diff
@@ -82,13 +82,15 @@ impl<Adapter: ConsensusAdapter + 'static> Engine<Proposal> for ConsensusEngine<A
             RLP_NULL
         };
 
+        let mut remove = false;
         let extra_data_hardfork = {
             let mut hardfork = self.node_info.hardfork_proposals.write().unwrap();
             match &*hardfork {
                 Some(v) => {
                     // remove invalid proposal
                     if v.block_number <= next_number {
                         hardfork.take();
```

Comment:
It should be removed after the execution of the transactions in this block is complete.